Include info in INVALID_DATA responses

jacobkeeler opened the pull request:
Adds error messages to `INVALID_DATA` responses when validation fails on a message to allow for simpler debugging.

Fixes #31.

jacobkeeler commented:
This PR will require some changes since method signature changes were made without deprecation

LuxoftAKutsan commented on src/components/formatters/include/formatters/CFormatterJsonBase.h:
```diff
@@ -99,7 +99,8 @@ class CFormatterJsonBase {
    */
   static void jsonValueToObj(
```
@jacobkeeler you added new parameter. I think that description should be updated

LuxoftAKutsan commented on src/components/smart_objects/include/smart_objects/default_shema_item.h:
```diff
@@ -98,9 +99,20 @@ CDefaultSchemaItem<Type>::CDefaultSchemaItem(const ParameterType& DefaultValue)
+                                                 std::string& errorMessage) {
+  if (getSmartType() != Object.getType()) {
+    if (!Object.getKey().empty()) {
+      errorMessage.assign("Validation failed for \"" + Object.getKey() +
```
@jacobkeeler not sure that initial error message should be replaced. Or is in output parameter?

LuxoftAKutsan commented on src/components/smart_objects/include/smart_objects/default_shema_item.h:
```diff
@@ -98,9 +99,20 @@ CDefaultSchemaItem<Type>::CDefaultSchemaItem(const ParameterType& DefaultValue)
+      errorMessage.assign("Validation failed for \"" + Object.getKey() +
+                          "\". ");
+    }
+    errorMessage += "Incorrect type, expected: " +
```
@jacobkeeler if errormessage - output parameter why you append to it? Why not replacing?

LuxoftAKutsan commented on src/components/smart_objects/include/smart_objects/enum_schema_item.h:
```diff
@@ -69,7 +69,8 @@ class TEnumSchemaItem : public CDefaultSchemaItem<EnumType> {
    * @return NsSmartObjects::Errors::eType
    **/
-  Errors::eType validate(const SmartObject& Object) OVERRIDE;
+  Errors::eType validate(const SmartObject& Object,
```
@jacobkeeler you added parameter. You should fix description

LuxoftAKutsan commented on src/components/smart_objects/include/smart_objects/number_schema_item.h:
```diff
@@ -148,10 +159,26 @@ Errors::eType TNumberSchemaItem<NumberType>::validate(
+    std::stringstream stream;
+    stream << "Value too large, got: " << value
+           << ", maximum allowed: " << rangeLimit;
+    errorMessage += stream.str();
```
@jacobkeeler looks like errorMessage used something like a recursive validation report? std::string is not best container for such purpose. We have such validation type in policy table, you can do something the same.

LuxoftAKutsan commented on src/components/smart_objects/include/smart_objects/smart_schema.h:
```diff
@@ -67,7 +67,8 @@ class CSmartSchema FINAL {
    * @return Result of validation.
    */
-  Errors::eType validate(const SmartObject& Object) const;
+  Errors::eType validate(const SmartObject& Object,
```
@jacobkeeler I would make validate function OVERRIDE everywhere

LuxoftAKutsan commented on src/components/smart_objects/src/smart_object.cc:
```diff
@@ -838,6 +845,15 @@ SmartType SmartObject::getType() const {
 }
 
+void SmartObject::setKey(const std::string& NewKey) {
+  delete m_key;
```
@jacobkeeler why key is stored by pointer?

LuxoftAKutsan commented on src/components/application_manager/src/commands/mobile/system_request.cc:
```diff
@@ -559,7 +559,7 @@ void SystemRequest::Run() {
     }
 
-    CFormatterJsonBase::jsonValueToObj(root, sm_object);
+    CFormatterJsonBase::jsonValueToObj(root, sm_object, "");
```
@jacobkeeler may be it would be better to use default argument? Or overloaded function? Mandatory explicit setting key it confusing.

jacobkeeler commented:
@LuxoftAKutsan I have addressed all of your comments, this PR now uses the ValidationReport class to report errors in SmartObject validation instead of a simple string reference.
